### PR TITLE
feat(framework): pytorch

### PR DESCRIPTION
**Pull request by bojiang**
## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/frameworks/sklearn.py [hunk 1]
```diff
@@ -149,7 +149,14 @@ def save_model(
     )
 
     if signatures is None:
-        signatures = {"predict": {"batchable": False}}
+        signatures = {
```
Comment (parano) [inline bentoml/_internal/frameworks/sklearn.py:152]: revert this change? I think it's nicer to print a short info log that shows the default

File: bentoml/_internal/frameworks/sklearn.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 from bentoml.exceptions import NotFound
 from bentoml.exceptions import BentoMLException
 from bentoml.exceptions import MissingDependencyException
+from bentoml._internal.models.model import ModelSignaturesType
```
Comment (parano) [inline bentoml/_internal/frameworks/sklearn.py:14]: @sauyon  @larme let's use this type for `signature` parameter in all `save_model` methods?

File: bentoml/_internal/frameworks/pytorch_lightning.py [hunk 3]
```diff
@@ -63,32 +72,35 @@ def load(
...
         )
     weight_file = bentoml_model.path_of(f"{SAVE_NAMESPACE}{PT_EXT}")
-    model_format = bentoml_model.info.context.get("model_format")
-    # backward compatibility
-    if not model_format:
-        model_format = "pytorch_lightning:v1"
+    # model_format = bentoml_model.info.context.get("model_format")
+    # # backward compatibility
+    # if not model_format:
+    # model_format = "pytorch_lightning:v1"
```
Comment (parano) [inline bentoml/_internal/frameworks/pytorch_lightning.py:86]: remove this part?

File: bentoml/_internal/frameworks/pytorch.py [hunk 4]
```diff
@@ -100,8 +115,6 @@ def save(
             e.g. a tokenizer instance, preprocessor function, model configuration json
         metadata (:code:`Dict[str, Any]`, `optional`,  default to :code:`None`):
             Custom metadata for given model.
-        model_store (:mod:`~bentoml._internal.models.store.ModelStore`, default to :mod:`BentoMLContainer.model_store`):
-            BentoML modelstore, provided by DI Container.
 
     Returns:
         :obj:`~bentoml.Tag`: A :obj:`tag` with a format `name:version` where `name` is the user-defined model's name, and a generated `version` by BentoML.
```
Comment (parano) [inline bentoml/_internal/frameworks/pytorch.py:120]: update the docstring below?

File: bentoml/_internal/frameworks/torchscript.py [hunk 5]
```diff
@@ -103,104 +114,98 @@ def save(
...
+            def configure_optimizers(self):
+                return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)
 
-        tag = bentoml.torchscript.save("resnet50", resnet50)
-    """  # noqa
-    context: t.Dict[str, t.Any] = {
-        "framework_name": "torch",
-        "pip_dependencies": [f"torch=={get_pkg_version('torch')}"],
-    }
+        tag = bentoml.pytorch_lightning.save("lit_classifier", LitClassifier())
```
Comment (parano) [inline bentoml/_internal/frameworks/torchscript.py:157]:
```suggestion
        tag = bentoml.pytorch_lightning.save_model("lit_classifier", LitClassifier())
```

Comment (parano) [reply bentoml/_internal/frameworks/torchscript.py:157]: also, `signature` is required for this example

File: bentoml/_internal/frameworks/torchscript.py [hunk 6]
```diff
@@ -1,97 +1,108 @@
...
+    _include_pytorch_lightning_version: bool = False,
 ) -> Tag:
     """
     Save a model instance to BentoML modelstore.
 
     Args:
         name (:code:`str`):
             Name for given model instance. This should pass Python identifier check.
-        model (:code:`torch.jit.ScriptModule`):
+        model (`pl.LightningModule`):
```
Comment (parano) [inline bentoml/_internal/frameworks/torchscript.py:98]: this should be `torch.jit.ScriptModule` right?

Comment (parano) [reply bentoml/_internal/frameworks/torchscript.py:98]: also make sure to use .rst format, e.g.  
```
model (:code:`torch.jit.ScriptModule`)
```

File: bentoml/_internal/frameworks/torchscript.py [hunk 7]
```diff
@@ -103,104 +114,98 @@ def save(
...
+        }
+    else:
+        framework_versions = {"torch": get_pkg_version("torch")}
+
+    context: ModelContext = ModelContext(
+        framework_name="torchscript",
+        framework_versions=framework_versions,
+    )
+
+    if signatures is None:
```
Comment (parano) [inline bentoml/_internal/frameworks/torchscript.py:173]: does it make sense also to apply a default signature for all PyTorch models, with { "__call__": { batchable: True }}` ?

File: bentoml/_internal/frameworks/common/pytorch.py [hunk 8]
```diff
@@ -41,141 +41,156 @@
...
 class PyTorchTensorContainer(DataContainer[torch.Tensor, torch.Tensor]):
     @classmethod
-    def singles_to_batch(cls, singles, batch_axis=0):
-        return torch.stack(singles, dim=batch_axis)
+    def batches_to_batch(
+        cls,
+        batches: tuple[torch.Tensor] | list[torch.Tensor],
+        batch_dim: int = 0,
+    ) -> t.Tuple[torch.Tensor, list[int]]:
+        # numpy.concatenate may consume lots of memory, need optimization later
```
Comment (sauyon) [inline bentoml/_internal/frameworks/common/pytorch.py:107]: Does this still apply to tensors?

File: bentoml/_internal/frameworks/common/pytorch.py [hunk 9]
```diff
@@ -41,141 +41,156 @@
...
+        indices: t.Sequence[int],
+        batch_dim: int = 0,
+    ) -> t.List[torch.Tensor]:
+        sizes = [indices[i] - indices[i - 1] for i in range(1, len(indices))]
+        return torch.split(batch, sizes, axis=batch_dim)
 
     @classmethod
     @inject
-    def single_to_payload(
+    def to_payload(  # pylint: disable=arguments-differ
```
Comment (sauyon) [inline bentoml/_internal/frameworks/common/pytorch.py:130]: `to_payload` should take a `batch_dim` now and set payload length for validation on the service app side.

File: bentoml/_internal/frameworks/torchscript.py [hunk 10]
```diff
@@ -1,97 +1,108 @@
...
+
+
+try:
+    import torch
+except ImportError:  # pragma: no cover
+    raise MissingDependencyException(_PL_IMPORT_ERROR)
 
-MODULE_NAME = "bentoml.torchscript"
+MODULE_NAME = "bentoml.pytorch_lightning"
+MODEL_FILENAME = "savd_model.pt"
```
Comment (sauyon) [inline bentoml/_internal/frameworks/torchscript.py:33]:
```suggestion
MODEL_FILENAME = "saved_model.pt"
```

File: bentoml/_internal/frameworks/torchscript.py (hunk 7 above)
Comment (sauyon) [reply bentoml/_internal/frameworks/torchscript.py:173]: Should we be enabling `batchable` by default?

File: bentoml/_internal/models/model.py [hunk 11]
```diff
@@ -42,8 +42,8 @@
     from ..types import PathType
 
     class ModelSignatureDict(t.TypedDict, total=False):
-        batch_dim: tuple[int, int]
-        batchable: bool
+        batch_dim: tuple[int, int] | int | None
+        batchable: bool | None
         input_spec: tuple[AnyType] | AnyType | None
         output_spec: AnyType | None
```
Comment (sauyon) [inline bentoml/_internal/models/model.py:48]: Does `total=False` not do the optional bit by default?

**Review by sauyon — COMMENTED** [review]
LGTM minus minor comments.

File: bentoml/_internal/frameworks/torchscript.py (hunk 7 above)
Comment (parano) [reply bentoml/_internal/frameworks/torchscript.py:173]: @sauyon I think it make sense for frameworks like PyTorch and TF, although I understand it might be better to keep the default behavior consistent and ask users to set this explicitly.

Btw we currently have batchable=True as default in the Keras module

File: bentoml/_internal/frameworks/pytorch.py [hunk 12]
```diff
@@ -1,48 +1,53 @@
...
-from .common.pytorch import PyTorchTensorContainer  # pylint: disable=unused-import
-from ..configuration.containers import BentoMLContainer
+from ..models.model import ModelContext
+from ..models.model import ModelSignaturesType
+from .common.pytorch import torch  # type: ignore
+from .common.pytorch import PyTorchTensorContainer
 
-if TYPE_CHECKING:
-    from ..models import ModelStore
+__all__ = ["load_model", "get_runnable", "get", "PyTorchTensorContainer"]
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/pytorch.py:22]: why do we need this here? we should only set `__all__` at the top level

File: bentoml/_internal/frameworks/torchscript.py (hunk 7 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/torchscript.py:173]: @parano @sauyon Pytorch modules are not using `__call__` as the default inference method, nor in its official guide, thus it doesn't make sense, I believe.

File: bentoml/_internal/frameworks/pytorch.py [hunk 13]
```diff
@@ -53,37 +58,29 @@ def load(
...
-            model: "torch.nn.Module" = cloudpickle.load(file).to(device_id)
-    elif model_format == "torch.save:v1":
-        with Path(weight_file).open("rb") as file:
-            model: "torch.nn.Module" = torch.load(file, map_location=device_id)
-    else:
-        raise BentoMLException(f"Unknown model format {model_format}")
 
+    weight_file = bentoml_model.path_of(MODEL_FILENAME)
+    with Path(weight_file).open("rb") as file:
+        model: "torch.nn.Module" = torch.load(file, map_location=device_id)
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/pytorch.py:72]: should we keep `model_format` behaviour like before?

File: bentoml/_internal/frameworks/pytorch_lightning.py [hunk 14]
```diff
@@ -156,72 +103,24 @@ def configure_optimizers(self):
...
-        name=name,
-        partial_kwargs=partial_kwargs,
+    return script_save_model(
+        name,
+        script_module,
+        signatures=signatures,
+        labels=labels,
+        custom_objects=custom_objects,
+        metadata=metadata,
+        _include_pytorch_lightning_version=True,
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/pytorch_lightning.py:125]: hmm this seems to be like an internal arg from torchscript. 🤔

File: bentoml/_internal/frameworks/torchscript.py (hunk 5 above)
Comment (aarnphm) [inline bentoml/_internal/frameworks/torchscript.py:157]: Since this is the torchscript model, We shouldn't show the pytorch lightning examples. PyTorch lightning example should be under pytorch lightning instead. Even though pytorch lightning use torchscript under the hood, I think it is still better to have the separation and let user know about that.

File: bentoml/_internal/models/model.py [hunk 15]
```diff
@@ -490,12 +488,19 @@ def __init__(
         api_version: str | None = None,
         creation_time: datetime | None = None,
     ):
-        ...
-
-    def __attrs_post_init__(self):
-        object.__setattr__(self, "name", self.tag.name)
-        object.__setattr__(self, "version", self.tag.version)
-
```
Comment (aarnphm) [inline bentoml/_internal/models/model.py:498]: should we use `field_transformer` instead here?

File: bentoml/_internal/frameworks/pytorch.py (hunk 13 above)
Comment (parano) [reply bentoml/_internal/frameworks/pytorch.py:72]: What is that for?

File: bentoml/_internal/models/model.py (hunk 15 above)
Comment (parano) [reply bentoml/_internal/models/model.py:498]: The current approach is ok

File: bentoml/_internal/frameworks/torchscript.py (hunk 7 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/torchscript.py:173]: Let me check about the original usage or __call__ in its doc

File: bentoml/_internal/frameworks/pytorch.py (hunk 13 above)
Comment (aarnphm) [reply bentoml/_internal/frameworks/pytorch.py:72]: for compatibility i believe right @larme ?

Comment (larme) [reply bentoml/_internal/frameworks/pytorch.py:71]: I think we can drop model format backward compatibility before 1.0 release to simplify the code. But keep backward compatibility after 1.0 release

File: bentoml/_internal/frameworks/torchscript.py (hunk 7 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/torchscript.py:137]: Just updated. will use __call__ as the default method. But I don't think we should have batchable=True for models, which may cause unwanted batching behavior

File: bentoml/_internal/frameworks/pytorch.py (hunk 12 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/pytorch.py]: This is the standard way to get linter to stop complaining about A `imported but not used`

File: bentoml/_internal/frameworks/pytorch_lightning.py (hunk 14 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/pytorch_lightning.py:124]: We don't want to expose this arg to users, I think.

Comment (aarnphm) [reply bentoml/_internal/frameworks/pytorch_lightning.py:124]: agree.

File: bentoml/_internal/models/model.py (hunk 15 above)
Comment (bojiang) [reply bentoml/_internal/models/model.py:498]: the "__attrs_init__" is the recommended way in case the initialization parameter type of the class is not the same as the property type. Naturally friendly to typing tools.

File: bentoml/_internal/frameworks/pytorch.py (hunk 4 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/pytorch.py:105]: done

File: bentoml/_internal/frameworks/common/pytorch.py (hunk 9 above)
Comment (bojiang) [reply bentoml/_internal/frameworks/common/pytorch.py:125]: done

File: bentoml/_internal/models/model.py (hunk 11 above)
Comment (bojiang) [reply bentoml/_internal/models/model.py]: TIL
